platform/mbed_retarget.cpp

```diff
@@ -205,82 +364,100 @@ extern "C" FILEHANDLE PREFIX(_open)(const char* name, int openmode) {
     }
-    #endif
-
-    // find the first empty slot in filehandles
-    filehandle_mutex->lock();
-    unsigned int fh_i;
-    for (fh_i = 0; fh_i < sizeof(filehandles)/sizeof(*filehandles); fh_i++) {
-    	/* Take a next free filehandle slot available. */
-        if (filehandles[fh_i] == NULL) break;
+#endif
+#ifndef __IAR_SYSTEMS_ICC__
+    /* FILENAME: "@(integer)" gives an already-allocated descriptor */
+    if (name[0] == '@') {
+        int fd;
+        memcpy(&fd, name + 1, sizeof fd);
+        return fd;
     }
-    if (fh_i >= sizeof(filehandles)/sizeof(*filehandles)) {
-        /* Too many file handles have been opened */
-        errno = EMFILE;
-        filehandle_mutex->unlock();
-        return -1;
+#endif
+    return open(name, openflags_to_posix(openflags));
+}
+
+extern "C" int open(const char* name, int oflag, ...) {
+    int fh_i = reserve_filehandle();
+    if (fh_i < 0) {
+        return fh_i;
     }
-    filehandles[fh_i] = (FileHandle*)FILE_HANDLE_RESERVED;
-    filehandle_mutex->unlock();
 
     FileHandle *res = NULL;
+    FilePath path(name);
 
-    /* FILENAME: ":(pointer)" describes a FileHandle* */
-    if (name[0] == ':') {
-        void *p;
-        memcpy(&p, name + 1, sizeof(p));
-        res = (FileHandle*)p;
+    if (!path.exists()) {
+        /* The first part of the filename (between first 2 '/') is not a
+         * registered mount point in the namespace.
+         */
+        return handle_open_errors(-ENODEV, fh_i);
+    }
 
-    /* FILENAME: "/file_system/file_name" */
+    if (path.isFile()) {
+        res = path.file();
     } else {
-        FilePath path(name);
-
-        if (!path.exists()) {
-            /* The first part of the filename (between first 2 '/') is not a
-             * registered mount point in the namespace.
-             */
+        FileSystemHandle *fs = path.fileSystem();
+        if (fs == NULL) {
             return handle_open_errors(-ENODEV, fh_i);
         }
-
-        if (path.isFile()) {
-            res = path.file();
-        } else {
-            FileSystemHandle *fs = path.fileSystem();
-            if (fs == NULL) {
-                return handle_open_errors(-ENODEV, fh_i);
-            }
-            int posix_mode = openmode_to_posix(openmode);
-            int err = fs->open(&res, path.fileName(), posix_mode);
-            if (err) {
-                return handle_open_errors(err, fh_i);
-            }
+        int err = fs->open(&res, path.fileName(), oflag);
+        if (err) {
+            return handle_open_errors(err, fh_i);
         }
     }
 
     filehandles[fh_i] = res;
+    stdio_in_prev[fh_i] = 0;
+    stdio_out_prev[fh_i] = 0;
+
+    return fh_i;
+}
+
+int mbed::bind_to_fd(FileHandle *fh) {
+    int fh_i = reserve_filehandle();
+    if (fh_i < 0) {
+        return fh_i;
+    }
+
+    filehandles[fh_i] = fh;
+    stdio_in_prev[fh_i] = 0;
+    stdio_out_prev[fh_i] = 0;
 
-    return fh_i + 3; // +3 as filehandles 0-2 are stdin/out/err
+    return 0;
```

Comment:
should not we return `fh_i` here?

Reply:
Um..., yes? Also where it's called in `fdopen(FileHandle *)`, that should be checking for it returning < 0.

Thanks for that. Guess you're running your tests on it?